CrossSectionAnalysis/cpp/CrossSectionCompute.cxx

```diff
@@ -0,0 +1,340 @@
+    os << indent << "inputSurfaceNode: " << this->inputSurfaceNode << "\n";
+    os << indent << "inputSegmentID: " << this->inputSegmentID << "\n";
+}
+
+// The surface polydata is constant. Create it once only.
+bool vtkCrossSectionCompute::SetInputSurfaceNode(vtkMRMLNode * inputSurface, const std::string& inputSegmentId)
+{
+  inputSurfaceNode = inputSurface;
+  inputSegmentID = inputSegmentId;
+  if (inputSurfaceNode->IsA("vtkMRMLSegmentationNode"))
+  {
+    vtkMRMLSegmentationNode * inputSegmentationNode = vtkMRMLSegmentationNode::SafeDownCast(inputSurfaceNode);
+    inputSegmentationNode->CreateClosedSurfaceRepresentation();
+    inputSegmentationNode->GetClosedSurfaceRepresentation(inputSegmentID, closedSurfacePolyData);
+  }
+  else if (inputSurfaceNode->IsA("vtkMRMLModelNode"))
+  {
+    vtkMRMLModelNode * inputModelNode = vtkMRMLModelNode::SafeDownCast(inputSurfaceNode);
+    closedSurfacePolyData->DeepCopy(inputModelNode->GetPolyData());
+  }
+  else
+  {
+    std::cout << "Invalid closed surface." << std::endl;
+    return false;
+  }
+  return true;
+}
+
+void vtkCrossSectionCompute::UpdateTable(vtkDoubleArray * crossSectionAreaArray, vtkDoubleArray * ceDiameterArray)
+{
+  if (inputCenterlineNode == NULL)
+  {     
+    std::cout << "Input centerline is NULL." << std::endl;
+    return;
+  }
+  if (inputSurfaceNode == NULL)
+  {     
+    std::cout << "Input surface is NULL." << std::endl;
+    return;
+  }
+  if (inputSegmentID.empty())
+  {     
+    std::cout << "Input segment ID is unknown." << std::endl;
+    return;
+  }
+  /*
+   * Divide the number of centerline points in equal blocks per thread.
+   * The last block will include the residual points also.
+   */
+  const unsigned int numberOfValues = crossSectionAreaArray->GetNumberOfValues();
+  unsigned int residual = numberOfValues % numberOfThreads;
+  unsigned int numberOfValuesPerBlock = (numberOfValues - residual) / numberOfThreads;
+  
+  std::vector<std::thread> threads;
+  std::vector<vtkSmartPointer<vtkDoubleArray>> bufferArrays;
+  
+  for (unsigned int i = 0; i < numberOfThreads; i++)
+  {
+    unsigned int startPointIndex = i * numberOfValuesPerBlock;
+    unsigned int endPointIndex = ((i + 1) * numberOfValuesPerBlock) - 1;
+    if (i == (numberOfThreads -1))
+    {
+        endPointIndex += residual;
+    }
+
+    /* 
+     * Give each thread a copy of the closed surface.
+     */
+    vtkSmartPointer<vtkPolyData> closedSurfacePolyDataCopy = vtkSmartPointer<vtkPolyData>::New();
+    closedSurfacePolyDataCopy->DeepCopy(closedSurfacePolyData.Get());
+    
+    /* 
+     * Give each thread a copy of the centerline.
+     */
+    vtkMRMLNode * inputCenterlineNodeCopy;
+    
+    if (inputCenterlineNode->IsA("vtkMRMLModelNode"))
+    {
+        // Give each thread a copy of each centerline.
+        vtkMRMLModelNode * inputModel = vtkMRMLModelNode::SafeDownCast(inputCenterlineNode);
```

Comment:
This is not a full copy. For example, the node is not added to the scene, the parent transform is not applied to the node, etc. Creating a full copy would be a quite expensive operation and it would be very difficult to ensure that no events are invoked accidentally from any of the MRML nodes, so I would only pass plain VTK objects to the threads, not MRML nodes.